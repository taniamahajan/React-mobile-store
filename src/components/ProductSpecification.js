import React, { Component } from "react";
import axios from "axios";
import { Jumbotron, Button, Row, Col } from "react-bootstrap";
import { connect } from "react-redux";
import { getCartData } from "../actions/products";

class ProductSpecification extends Component {
  constructor() {
    super();
    this.state = {
      product: {},
    };
  }
  description = [
    "Wi-Fi or cellular access to the internet",
    "A battery that powers the device for several hours",
    "A physical or onscreen keyboard for entering information",
    "Size and weight that allows it to be carried in one hand and manipulated with the other hand",
    "Touch-screen interface in almost all cases",
    "A virtual assistant, like Siri, Cortana or Google Assistant",
    "The ability to download data from the internet, including apps and books",
    "Wireless operation",
  ];
  cart = {};
  componentDidMount() {
    const id = window.location.pathname;
    this.props.getCartData();
    axios.get("http://localhost:3333/products" + id).then((res) => {
      this.setState({
        product: res.data,
      });
    });
  }

  addToCart() {
    if (localStorage.getItem("userId") && this.props.cart) {
      let cart = this.props.cart;
      let productIndex = cart.products.findIndex(
        (x) => x.productId === this.state.product.id
      );
      if (productIndex >= 0) {
        cart.products[productIndex].quantity += 1;
      } else {
        cart.products.push({
          productId: this.state.product.id,
          quantity: 1,
        });
      }
      axios
        .patch("http://localhost:3333/carts/" + cart.id, {
          products: cart.products,
        })
        .then((res) => {});
    } else {
      alert("You need to login to add to cart");
      this.props.history.push("/login");
    }
  }

  render() {
    return (
      <div>
        <Row>
          <Col>
            <img
              className="image-product"
              src={this.state.product.image}
              alt={this.state.product.name}
            />
          </Col>
          <Col>
            <h1>{this.state.product.name}</h1>

            <ul>
              {this.description.map((x, index) => {
                return <li key={index + 1}>{x}</li>;
              })}
            </ul>
            <p>
              <h2>
                $
                {this.state.product.price
                  ? this.state.product.price.toLocaleString("en-US", {
                      style: "currency",
                      currency: "USD",
                    })
                  : null}
              </h2>
            </p>
            <p>
              <Button variant="primary" onClick={() => this.addToCart()}>
                Add to cart
              </Button>
            </p>
          </Col>
        </Row>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    cart: state.cart,
  };
}
function mapDispatchToProps(dispatch) {
  return {
    getCartData: () => dispatch(getCartData(1)),
  };
}
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProductSpecification);
