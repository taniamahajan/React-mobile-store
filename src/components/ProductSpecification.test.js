import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import ProductSpecification from "./ProductSpecification";
const mockStore = configureStore([]);
describe("My Connected React-Redux Component", () => {
  let store;
  let component;

  beforeEach(() => {
    Object.defineProperty(window, "location", {
      value: {
        pathname: 1,
      },
    });
    store = mockStore({
      myState: "sample text",
    });
    Storage.prototype.getItem = jest.fn(() => "1");

    jest.spyOn(window, "alert").mockImplementation(() => {});
    store.dispatch = jest.fn();

    component = renderer.create(
      <Provider store={store}>
        <ProductSpecification />
      </Provider>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should dispatch an action on button click", () => {
    renderer.act(() => {});

    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
