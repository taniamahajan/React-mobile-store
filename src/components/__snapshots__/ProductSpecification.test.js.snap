// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component should render with given state from Redux store 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <img
        className="image-product"
      />
    </div>
    <div
      className="col"
    >
      <h1 />
      <ul>
        <li>
          Wi-Fi or cellular access to the internet
        </li>
        <li>
          A battery that powers the device for several hours
        </li>
        <li>
          A physical or onscreen keyboard for entering information
        </li>
        <li>
          Size and weight that allows it to be carried in one hand and manipulated with the other hand
        </li>
        <li>
          Touch-screen interface in almost all cases
        </li>
        <li>
          A virtual assistant, like Siri, Cortana or Google Assistant
        </li>
        <li>
          The ability to download data from the internet, including apps and books
        </li>
        <li>
          Wireless operation
        </li>
      </ul>
      <p>
        <h2>
          $
        </h2>
      </p>
      <p>
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Add to cart
        </button>
      </p>
    </div>
  </div>
</div>
`;
